lib/host/wasi/inode-linux.cpp

```diff
@@ -911,34 +914,56 @@ WasiExpect<INode> INode::sockOpen(__wasi_address_family_t AddressFamily,
   }
 }
 
-WasiExpect<void> INode::sockBind(__wasi_address_family_t AddressFamily,
-                                 Span<const uint8_t> Address,
-                                 uint16_t Port) noexcept {
-  Variant<sockaddr, sockaddr_in, sockaddr_in6> ServerAddr;
-  size_t Size;
+static std::vector<uint8_t>
+sockAddressAssignHelper(__wasi_address_family_t AddrFamily,
+                        const Span<const uint8_t> &Address, uint16_t Port) {
+  std::vector<uint8_t> Buffer;
+  // Do nothing if Address is not existing.
+  if (Address.size() == 0)
+    return {};
 
-  if (AddressFamily == __WASI_ADDRESS_FAMILY_INET4) {
-    auto &ServerAddr4 = ServerAddr.emplace<sockaddr_in>();
-    Size = sizeof(ServerAddr4);
+  if (AddrFamily == __WASI_ADDRESS_FAMILY_INET4) {
+    Buffer.resize(sizeof(sockaddr_in));
+    auto *ServerAddr4 = reinterpret_cast<struct sockaddr_in *>(Buffer.data());
 
-    ServerAddr4.sin_family = AF_INET;
-    ServerAddr4.sin_port = htons(Port);
+    ServerAddr4->sin_family = AF_INET;
+    ServerAddr4->sin_port = htons(Port);
     assuming(Address.size() >= sizeof(in_addr));
-    std::memcpy(&ServerAddr4.sin_addr, Address.data(), sizeof(in_addr));
-  } else if (AddressFamily == __WASI_ADDRESS_FAMILY_INET6) {
-    auto &ServerAddr6 = ServerAddr.emplace<sockaddr_in6>();
-    Size = sizeof(ServerAddr6);
-
-    ServerAddr6.sin6_family = AF_INET6;
-    ServerAddr6.sin6_port = htons(Port);
+    std::memcpy(&ServerAddr4->sin_addr, Address.data(), sizeof(in_addr));
+  } else if (AddrFamily == __WASI_ADDRESS_FAMILY_INET6) {
+    Buffer.resize(sizeof(sockaddr_in6));
+    auto *ServerAddr6 = reinterpret_cast<struct sockaddr_in6 *>(Buffer.data());
+
+    ServerAddr6->sin6_family = AF_INET6;
+    ServerAddr6->sin6_port = htons(Port);
+    ServerAddr6->sin6_flowinfo = 0;
     assuming(Address.size() >= sizeof(in6_addr));
-    std::memcpy(&ServerAddr6.sin6_addr, Address.data(), sizeof(in6_addr));
+    std::memcpy(&ServerAddr6->sin6_addr, Address.data(), sizeof(in6_addr));
+  } else if (AddrFamily == __WASI_ADDRESS_FAMILY_AF_UNIX) {
+    Buffer.resize(sizeof(sockaddr_un));
+    auto *ServerAddrUN = reinterpret_cast<struct sockaddr_un *>(Buffer.data());
+
+    ServerAddrUN->sun_family = AF_UNIX;
+    // TODO: check the spec of sockaddr_un::sun_path
+    assuming(Address.size() >= sizeof(sockaddr_un::sun_path));
+    std::memcpy(&ServerAddrUN->sun_path, Address.data(),
```

Comment:
Maybe we will need a configuration to set the allowlist of the binding `sun_path`. If there are no options for it, then users may access the UNIX sockets that are not in the pre-open list.

For a workaround, we can have one global option called `--allow-af-unix` to enable this. If users don't specify, then, we should disable this feature.